### PR TITLE
stress test: add stashed ops

### PR DESCRIPTION
**Pull request by wes-carlson**
## Description
Add option to generate and stash ops when closing the container in stress test. At the end of the run, it verifies that the expected number of ops have been submitted.

## Reviewer Guidance
1) Implementing TaskManager.applyStashedOp() would be part of offline phase 6. I've removed the error it throws since we sometimes get stashed TaskManager ops
2) I don't expect the changes to ConnectionManager to go in with this PR. I've made a separate PR here: #14997
3) There is a bug in BlobManager when a BlobAttach op submitted by the previous clientId is processed after loading with stashed ops. This deletes the pending blob entry, tripping an assert when the upload initiated on load finishes because there is no pending entry. I've avoided the assert here but this should probably be fixed in a separate PR.
4) The other BlobManager changes are mostly copied from #11927 and should probably go in an updated version of that PR before this is merged.

### CODE REVIEW

File: packages/test/test-service-load/src/loadTestDataStore.ts [hunk 1]
```diff
@@ -36,11 +36,24 @@ export interface ILoadTest {
 	run(config: IRunConfig, reset: boolean): Promise<boolean>;
 	detached(config: Omit<IRunConfig, "runId" | "profileName">): Promise<LoadTestDataStoreModel>;
 	getRuntime(): Promise<IFluidDataStoreRuntime>;
+	waitLoaded(): Promise<void>;
+	/**
+	 * Generate ops.
+	 * @param n - number of ops to generate
+	 */
+	generateChanges(n: number);
```
Comment (anthony-murphy) [inline packages/test/test-service-load/src/loadTestDataStore.ts:44]: do we actually need to generate ops specifically for stashed ops. usually for this tests we try to keep features isolated, as it is more about leveraging features together, than stressing a specific feature in a specific way.

File: packages/test/test-service-load/src/loadTestDataStore.ts [hunk 2]
```diff
@@ -36,11 +36,24 @@ export interface ILoadTest {
...
+	/**
+	 * Verify that the expected number of stashed ops have been applied to the counter
+	 * @param expected - expected number of stashed ops
+	 * @returns true if the counter matches the expected value
+	 */
+	verify(expected: number): boolean;
```
Comment (anthony-murphy) [inline packages/test/test-service-load/src/loadTestDataStore.ts:50]: i don't think we should add this. this test isn't meant to verify specific features.

File: packages/test/test-service-load/src/runner.ts [hunk 3]
```diff
@@ -398,24 +432,59 @@ function scheduleContainerClose(
...
+											// Synchronously generate changes and close. There should be zero chance these ops
+											// round-trip before closing.
+											test.generateChanges(count);
+											def.resolve({
+												pendingState:
+													container.closeAndGetPendingLocalState(),
+												count,
+											});
+										})
+										.catch((e) => console.error(e));
```
Comment (anthony-murphy) [inline packages/test/test-service-load/src/runner.ts:451]: log an error to telemetry, not the console, so these errors are captured.

File: packages/test/test-service-load/src/runner.ts [hunk 4]
```diff
@@ -398,24 +432,59 @@ function scheduleContainerClose(
...
+								const count = runConfig.random.integer(
+									stashedOpsMin,
+									stashedOpsMax,
+								);
+								if (count > 0) {
+									test.waitLoaded()
+										.then(() => {
+											// Synchronously generate changes and close. There should be zero chance these ops
+											// round-trip before closing.
+											test.generateChanges(count);
```
Comment (anthony-murphy) [inline packages/test/test-service-load/src/runner.ts:444]: this should be captured via a new mechanism, which schedules the container going offline.

Comment (anthony-murphy) [reply packages/test/test-service-load/src/runner.ts:444]: it looks like we already have a schedule that will go offline, i think that should be enough, we then just need to schedule closeAndGetPendingLocalState if the container has gone offline, and supports offline

File: packages/test/test-service-load/src/runner.ts [hunk 5]
```diff
@@ -444,6 +513,9 @@ function scheduleOffline(
 				const injectionTime = random.integer(offlineDelayMinMs, offlineDelayMaxMs);
 				await new Promise<void>((resolve) => setTimeout(resolve, injectionTime));
 
+				if (container.closed) {
```
Comment (anthony-murphy) [inline packages/test/test-service-load/src/runner.ts:516]: i think i would make changes below here by calling closeAndGetPendingState if the container has offline enabled for some set of clients

File: packages/test/test-service-load/src/runner.ts [hunk 6]
```diff
@@ -444,23 +452,29 @@ function scheduleOffline(
...
 
 				await new Promise<void>((resolve) => setTimeout(resolve, offlineTime));
-				if (!container.closed) {
-					ds.goOnline();
-					printStatus(runConfig, "going online!");
-					return schedule();
+				if (container.closed) {
+					return undefined;
 				}
+				if (runConfig.loaderConfig?.enableOfflineLoad === true) {
```
Comment (anthony-murphy) [inline packages/test/test-service-load/src/runner.ts:458]: should this be probablity based, so we don't always close when offline, only sometimes.

**Review by anthony-murphy — APPROVED** [review]

File: packages/test/test-service-load/src/testConfigFile.ts [hunk 7]
```diff
@@ -45,6 +46,12 @@ export interface ILoadTestConfig {
 			min: number;
 			max: number;
 		};
+		/**
+		 * Percentage of containers for which to stash pending changes when the offline period ends. Note
+		 * that this will only affect containers that have offline load enabled. A float between 0 and 1,
+		 * 1 meaning always stash.
+		 */
+		stashPercent?: number;
```
Comment (anthony-murphy) [inline packages/test/test-service-load/src/testConfigFile.ts:54]: i don't see this in the config?

File: packages/test/test-service-load/src/runner.ts [hunk 8]
```diff
@@ -410,32 +428,37 @@ function scheduleContainerClose(
...
 	dsf: FaultInjectionDocumentServiceFactory,
 	container: IContainer,
 	runConfig: IRunConfig,
 	offlineDelayMinMs: number,
 	offlineDelayMaxMs: number,
 	offlineDurationMinMs: number,
 	offlineDurationMaxMs: number,
-) {
-	new Promise<void>((resolve) => {
+	stashPercent = 1,
```
Comment (anthony-murphy) [inline packages/test/test-service-load/src/runner.ts:439]: this should come from config, not be fixed

**Comment by alexvy86** [comment]
@wes-carlson @anthony-murphy starting with the first build after this commit, we started seeing a significant number of "ODSP fetch error [404]" errors in stress tests for odsp (and odsp-df). Is this expected? Or do we think coincidental (it's been several days though...).

```
// Count of errors by error message for odsp
let startTime=ago(5d);
let endTime=now();
office_fluid_ffautomation_error
| where Event_Time > startTime and Event_Time < endTime
| where Data_branch in ('refs/heads/main')
| where Data_hostName == "@fluid-internal/test-service-load"
| where Data_driverType == "odsp" and Data_driverEndpointName =="odsp"
| extend Data_buildId = toint(Data_buildId)
| summarize
    errorCount=count()
    by Data_buildId, Data_error
| sort by Data_buildId asc
| render linechart with (xcolumn=Data_buildId, ycolumns=errorCount, series=Data_error)
```
image

(@jikim-msft FYI)

**Pull request merged**